### ISSUE TITLE
Command bins is not working properly

### ISSUE THREAD

**Issue by Eterna1** [pwndbg/pwndbg#424]
### Description

Hello, this commit [6a640e8](https://github.com/pwndbg/pwndbg/commit/6a640e8ff29d4bdb957008b552621d894557e291
) has broken down `bins` command.

### Steps to reproduce

Here is the simple test:

```C
#include <stdio.h>

void foo() {}

int main() {
    char *a = malloc(0x16);
    malloc(0x16);
    free(a);
    foo();
} 
```

Set beakpoint on `foo` function and when it has been hit, type `bins` command.
Here is the result of the command after this commit: [pastebin](https://pastebin.com/v5ETCFL1)
Here is the result of the command before this commit: 

```
pwndbg> bins
fastbins
0x20: 0x602000 ◂— 0x0
0x30: 0x0
0x40: 0x0
0x50: 0x0
0x60: 0x0
0x70: 0x0
0x80: 0x0
unsortedbin
all: 0x0
smallbins
empty
largebins
empty
```

### My setup

I've been using 2 different setups so this is not the case probably

**Comment by Eterna1** [comment]
Can this commit be just reverted ?

**Comment by disconnect3d** [comment]
@Eterna1 It could be, for a hotfix, if the issue happened on stable branch, but it appears only on dev (so, as a dev, one can revert it by themselves - if you really need a working dev version).

Lets aim for fixing this, as the idea of the changes is fine.

**Comment by disconnect3d** [comment]
@AlexAltea would you be willing to help with that?

**Comment by AlexAltea** [comment]
I'm not familiar with the `bins` command. What exactly is the issue here? According to your snippets it seems as if my commit caused `smallbins` and `largebins` to be non-empty and `unsortedbin` to contain a different value in `all`. I don't even grasp why would my patch change that behavior.

@disconnect3d If this is my commit's fault, then I'd definitely try to debug and fix it myself, starting from March 12th (I'm afraid that before that day I'll be too busy with RL-projects).

**Comment by disconnect3d** [comment]
@AlexAltea I am not sure if that's ur fault but it seems like so :s

I really need to prepare myself a setup for working with heap stuff...

**Comment by mak** [comment]
So the difference now is that `chain.get` pushes address before checking a `hard_stop` limit in the past was the other way around and since bins are initialized with their corresponding address  loop was terminated returning empty list

if current semantics of `chain.get` is preferable then easy fix will be to check value in bins before resolving a chain, or add a flag to `chain.get` that wont append first address before the loop

**Comment by veritas501** [comment]
try this patch for `get` in `pwndbg/chain.py`

```python
def get(address, limit=LIMIT, offset=0, hard_stop=None, hard_end=0):
    """
    Recursively dereferences an address. For bare metal, it will stop when the address is not in any of vmmap pages to avoid redundant dereference.

    Arguments:
        address(int): the first address to begin dereferencing
        limit(int): number of valid pointers
        offset(int): offset into the address to get the next pointer
        hard_stop(int): address to stop at
        hard_end: value to append when hard_stop is reached

    Returns:
        A list representing pointers of each ```address``` and reference
    """
    limit = int(limit)

    #result = [address]
    result = [] #patch 1
    for i in range(limit):
        # Don't follow cycles, except to stop at the second occurrence.
        if result.count(address) >= 2:
            break

        if hard_stop is not None and address == hard_stop:
            result.append(hard_end)
            break

        result.append(address)#patch 2
        try:
            address = address + offset

            # Avoid redundant dereferences in bare metal mode by checking
            # if address is in any of vmmap pages
            if not pwndbg.abi.linux and not pwndbg.vmmap.find(address):
                break

            address = int(pwndbg.memory.poi(pwndbg.typeinfo.ppvoid, address))
            address &= pwndbg.arch.ptrmask
            #result.append(address)  #patch 3
        except gdb.MemoryError:
            break

    return result
```

**Comment by disconnect3d** [comment]
@veritas501 Thx for posting the patch! In the future please use pull requests for that :P (I have missed this one somehow :/).

I am going to check whether chain/telescope functionality works fine after those changes and merge it if it works.

**Comment by disconnect3d** [comment]
@veritas501 After your proposed patch the `set dereference-limit 0` -> `context` crashes.

Anyway, I have applied a fix proposed by @mak - by addding an `include_start` keyword argument to the `chain.get`. This lets us prevent adding the start address to the chain inside *bins commands.

The PR can be found here: https://github.com/pwndbg/pwndbg/pull/450

As the problem has been fixed, I am closing this issue :).
